### ISSUE TITLE
Loudness according to Annex F of ECMA-74

### ISSUE THREAD

**Issue by mglesser** [Eomys/MoSQITo#13]
Thanks to @djcaminero work, we have a first draft of [ECMA-74](https://www.ecma-international.org/publications-and-standards/standards/ecma-74/) annex F hearing model to compute loudness in MOSQITO. However  the model doesn't pass the validation stage yet (see figure below with the loudness contour from ISO 226 in dotted line and the loudness contour from MOSQITO hearing model in solid lines).

![valid_loudness_ecma](https://user-images.githubusercontent.com/60893296/120209654-38600380-c22f-11eb-98e7-cc239bda9cf9.png)

Any help from the community to review @djcaminero [code ](https://github.com/djcaminero/MoSQITo)will be welcome. You can reproduce the figure above by running mosqito/validation/hearing_model/hearing_model_validation.py script. That's a good starting point for a review. 

If you would like to contribute but are not familiar with github, please contact me. I will be happy to help.

**Comment by mglesser** [comment]
I'm begining a step by step validation of the implementation. Step 1 is the outer and middle/inner ear filtering (section F 3.2 of the ECMA). By stopping line 229 of  \mosqito\functions\hearing_model\comp_loudness.py and plotting the frequency response of the filter:
```
w, h = signal.sosfreqz(sos_ear, worN=1500, fs=fs)  
db = 20*np.log10(np.maximum(np.abs(h), 1e-5))  
plt.semilogx(w, db)
```
I get a response compliant to the standard, see figure below (ECMA-74 on the left,  MOSQITO on the right)
![ecma-74_fig_F3_ear_filter_cmp](https://user-images.githubusercontent.com/60893296/120796856-9e080480-c53b-11eb-9ad0-6f48edec8acc.png)

**Comment by mglesser** [comment]
Second step is the gammatone filters design. I have created a new 'gammatone' function from 'afb_coefficients' and part of 'afb_ppal_parameters' the objective is to be able to easily compare MOSQITO implementation of the gammatone filter with the one of scipy. The function is not used in the main code yet, but it should be eventually. 

The filters from mosqito match the one from scipy enough to consider that it is, at least, not totally wrong considering that it is not necessarily the same auditory model that is implemented (run mosqito/validations/hearing_model/valid_gammatone.py to get the plots):
![gammatone_400Hz](https://user-images.githubusercontent.com/60893296/120829776-141d6300-c55e-11eb-8c06-c48f1bddbc5e.png)
![gammatone_4000Hz](https://user-images.githubusercontent.com/60893296/120829815-20a1bb80-c55e-11eb-962d-f2ed4ab7b7bd.png)

**Comment by mglesser** [comment]
I checked the validation script and realized that it was not giving a correct evaluation of the implementation. I've programmed an [alternative validation script](https://github.com/djcaminero/MoSQITo/blob/master/mosqito/validations/hearing_model/hearing_model_validation_alt.py). The principle is the following : for each frequency, the Loudness is calculated for 4 tones with an SPL amplitude determined by the isophone curves 20, 40, 60 and 80. A satisfying implementation should return values in phon close to 20, 40, 60 and 80 respectively. The result of the current implementation is the following (theoretical isophon in dotted lines, results from MOSQITO in solid lines):
![valid_loudness_ecma_alt](https://user-images.githubusercontent.com/60893296/122580787-9de92800-d056-11eb-9e7e-f85e50b27130.png)

**Comment by kmh1201** [comment]
I changed variable _centre_freq_array_ in the call of function `annex_f_validation()` in `comp_loudness()` to variable _centre_freq_ for working diagnostics.

**Comment by mglesser** [comment]
I've been reviewing the whole code. I took this opportunity to implement an alternative version of the [comp_loudness function](https://github.com/djcaminero/MoSQITo/blob/master/mosqito/functions/hearing_model/comp_loudness_alt.py). Some of the steps are simplified to allow for easyer debugging.

**Comment by mglesser** [comment]
I noticed that if I do not apply the Outer and middle/inner ear filtering (section 5.1.2 of the ECMA-418-2), I get a overall trend of the hearing model that is closer to the theoretical equal loudness contours. It suggests that the filter is not applied correctly to the signal.
 
![image](https://user-images.githubusercontent.com/60893296/127649012-658c918e-6744-42e8-b342-2d03dd6bf66f.png)

**Comment by mglesser** [comment]
I've finally spotted two errors in the "outer and midle ear filtering" section of the code.

1. The sosfiltfilt function was used to apply the filter. 
```Python
signal_filtered = sp.signal.sosfiltfilt(sos_ear, signal, axis=0)
```
As stated in the doc : "This function applies a linear digital filter twice, once forward and once backwards. The combined filter has zero phase and a filter order twice that of the original." The filter order was then twice too high.

2. The filter predict the response of the ear. In order to "compensate" this response and go from physical amplitude in dB to physiological amplitude in sone, one should apply the inverse filter:
```Python
b, a = sp.signal.sos2tf(sos_ear)
signal_filtered = sp.signal.lfilter(a, b, signal, axis=0)
```

We are now getting dangerously close to the theoretical equal loudness contours. Next mistakes are to be search in a portion of the code that affect differently the low and high amplitudes. I would guess the "Non-linerarity" section. 
![image](https://user-images.githubusercontent.com/60893296/127673562-960d98c5-8f00-433d-916e-ccb3acc71c14.png)

**Comment by kmh1201** [comment]
> 
> 
> I've finally spotted ~two~ one errors in the "outer and midle ear filtering" section of the code.
> 
>     1. The sosfiltfilt function was used to apply the filter.
> 
> 
> ```python
> signal_filtered = sp.signal.sosfiltfilt(sos_ear, signal, axis=0)
> ```
> 
> As stated in the doc : "This function applies a linear digital filter twice, once forward and once backwards. The combined filter has zero phase and a filter order twice that of the original." The filter order was then twice too high.
> 
> ~2. The filter predict the response of the ear. In order to "compensate" this response and go from physical amplitude in dB to physiological amplitude in sone, one should apply the inverse filter:~
> 
> ~We are now getting dangerously close to the theoretical equal loudness contours. Next mistakes are to be search in a portion of the code that affect differently the low and high amplitudes. I would guess the "Non-linerarity" section.~
> 
> ![image](https://user-images.githubusercontent.com/60893296/129394761-73583543-07ae-4fce-b78e-4c94b0f9a64c.png)

**Comment by mglesser** [comment]
Solved by PR #33
